test/cpp/end2end/async_end2end_test.cc

```diff
@@ -807,7 +807,11 @@ class AsyncEnd2endServerTryCancelTest : public AsyncEnd2endTest {
     EXPECT_FALSE(context->IsCancelled());
     context->TryCancel();
     gpr_log(GPR_INFO, "Server called TryCancel()");
-    EXPECT_TRUE(context->IsCancelled());
+    while (!context->IsCancelled()) {
```

Comment:
I was initially thinking the same but when I reasoned about it some more, I felt that this is actually the correct thing to do.
These tests were not intended to test the async'Ness of TryCancel () API; Instead, they are supposed to test that the RPC is cancelled properly (when using the C++ async APIs)  when the actual cancellation happens at various points of interest during processing the RPCs (i.e. see  CANCEL_BEFORE / CANCEL_AFTER and CANCEL_DURING_PROCESSING flags)

If we let TryCancel() be async, it becomes harder to test CANCEL_BEFORE (and also CANCEL_AFTER scenarios).. 

Sure, it *feels* like we are making the test  deterministic by snchronously waiting for TryCancel() to complete; but that's not the case; The CANCEL_DURING_PROCESSING cases are non deterministic (and are semantically equivalent to not Synchronously waiting for TryCancel() to actually cancel the RPC).